misk/src/main/kotlin/misk/hash/ConsistentHashRing.kt

```diff
@@ -0,0 +1,44 @@
+package misk.hash
+
+import com.google.common.hash.HashFunction
+import com.google.common.hash.Hashing
+
+/** A [ConsistentHashRing] maps strings to values based on a consistent hash */
+class ConsistentHashRing<T>(
+  members: Map<String, T>,
+  private val hashFn: HashFunction = Hashing.murmur3_32(),
+  private val vnodesCount: Int = 16
+) {
+  private val vnodes: IntArray
+  private val vnodesToMembers: Map<Int, T>
+
+  init {
+    val vnodesBuilder = mutableListOf<Int>()
+    val vnodesToMembers = mutableMapOf<Int, T>()
+
+    // Hash each member up to the vnode count to occupy slots in the ring. To resolve
+    // a resource, we find the vnode that is closest to the resources hash, then find
+    // the member corresponding to that vnode
+    members.forEach { (memberKey, member) ->
+      (0 until vnodesCount).forEach { replica ->
```

Comment:
This inner loop places a member on `vnodesCount` nodes, correct? Is this for redundancy?

Reply:
It's to distribute resources more evenly across nodes, and reduce the amount of resource shifting that occurs when the ring changes